### PR TITLE
Add better support for the java programming language

### PR DESCRIPTION
**Pull request by michaelpog**
* Add all types for java
* Add more operators for java
* Add modifiers for java
* Add shortcuts for methods
* Add arrays

### CODE REVIEW

**Comment by pokey** [comment]
This looks good, tho I'd be inclined to rebase it onto #586 and figure out how it will interact with the new more generic type stuff

**Comment by michaelpog** [comment]
Thank you for looking into it @pokey . I'm not sure how it's related to #586.
Would you mind clarifying please?

**Comment by pokey** [comment]
#586 uses `code_type` instead of a Python-specific list, and introduces actions for inserting types that can be overridden by specific programming languages. It would be good to try to do the same thing to a certain extent for Java, eg define the `code_type` list for the Java context and override the insert type / return type actions rather than having a new talon command. Don't need everything to be generic, but at least the basic type insertion would be good to use this new mechanism. Make sense?

**Comment by michaelpog** [comment]
Let me know if this is enough. I'll let the renaming from code_types to code_types and everything else, be taken care of by the other pr.

File: lang/java/java.talon [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+[state] {user.java_other_modifiers}: 
+    insert(user.java_other_modifiers)
+    key("space") 
+
+op array:
+    user.code_operator_subscript()
+
+op new:
+    insert("new ")
+op plus plus:
```
Comment (splondike) [inline lang/java/java.talon:53]: I notice you've also added 'op increment'. Should we just remove 'op plus plus' here?

File: lang/java/java.talon [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    insert("++")
+
+code_block_comment
```
Comment (splondike) [inline lang/java/java.talon:56]: Leftover line.

File: lang/java/java.talon [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+# Arrays
+type {user.code_types} array:
+    insert(user.code_types)
+    user.code_operator_subscript()
+
+type ({user.java_common_classes}|{user.java_generic_data_structures}) array:
+    insert(java_common_classes or java_generic_data_structures)
+    user.code_operator_subscript()    
+
+[state] {user.java_access_modifiers}:
```
Comment (splondike) [inline lang/java/java.talon:40]: C#, Python, and javascript all use 'pro static funky' type things to make functions (and assume a 'void' return type in the case of C#). Maybe the java stuff should work the same way as them? There's no direct support for class access modifiers in the C# implementation either.

I guess the difference would be between saying "public static boxed integer camel my fancy method" -> "public static Integer myFancyMethod|" and "pub static my fancy method" -> "public static void myFancyMethod(|)"and then having to go back and change the void to an integer in the second case.

File: lang/java/java.talon [hunk 4]
```diff
@@ -0,0 +1,56 @@
+mode: user.java
+mode: user.auto_lang
+and code.language: java
+-
+tag(): user.code_operators
+tag(): user.code_comment
+tag(): user.code_block_comment
+tag(): user.code_generic
+
+action(user.code_state_import):
```
Comment (splondike) [inline lang/java/java.talon:10]: Do this .py style, implementing actions in .talon files is deprecated.

File: lang/java/java.talon [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    insert("import ")
+
+# Types Commands
+type {user.code_types}: 
+    insert(code_types)
+    key("space")
+
+type ({user.java_common_classes}|{user.java_generic_data_structures}):
```
Comment (splondike) [inline lang/java/java.talon:18]: I'd say maybe swap your .py stuff to put all the unboxed stuff into code_types along with the common classes and generic data structures, and then you could just fall back to the 'type *' command defined in programming.talon.

**Review by splondike — COMMENTED** [review]
`user.code_types` has been renamed to `user.code_type` as part of the referenced PR (#586), so this PR doesn't actually work properly when merged in (e.g. "type short" doesn't insert "Short"). I've put in some other comments as well for your consideration.

**Comment by michaelpog** [comment]
Comments are addressed

**Comment by splondike** [comment]
Looks good, thanks for the updates!

**Pull request merged**